### DISCUSSION TITLE
Ability to create generic types with default containing type?

### DISCUSSION THREAD

**Discussion by joshua-mng** [dotnet/csharplang#6627] [Language Ideas]
Take a look at the following piece of code:

```
class HomeController
{
    public Test(ILogger<HomeController> logger, IStringLocalizer<HomeController> localizer)
    {
    }
}
```

This usage of injecting generic interface with containing type is so common specially in asp.net core.

Is there any way to simplify this injection, like ILogger<this>, IStringLocalizer<this> or other forms.

This maybe oversimplification (since refactoring will correct usages of class names anyway), but it's overwhelming to type long class names over and over again. 

This maybe a great addition for C# language. Please let me know what you think!

Thanks

**Comment by HaloFour** [comment]
I'm going to assume that the only reason you've made `ILogger` and `IStringLocalizer` generic is so that you can specify the injectee as the generic type argument.  Why not utilize the facilities of whatever dependency injection library that you're using in order to capture said injectee during injection?  Then those interfaces wouldn't need to be generic at all and you wouldn't need some weird new syntax to "alias" the declaring class as a generic type argument.

**Comment by dsaf** [comment]
Problem solved (unless of course you meant "infer" rather than "default"):

``` csharp
public interface IHomeControllerLogger : ILogger<HomeController> {}
```

Alternatively dotnet/roslyn#7451.

**Comment by idilov** [comment]
I believe Joshua meant "infer". I'd like to do exactly what HaloFour suggested but I cannot find how to do it.
Please post a link to a resource explaining how to capture injectee using standard asp.net core di.

**Comment by jnm2** [comment]
This is like https://github.com/dotnet/csharplang/issues/6300 but for type parameters rather than method parameters.

**Comment by TahirAhmadov** [comment]
I actually think this can be done with a `self` constraint and required properties:
```cs
abstract class ControllerBase<TSelf> where TSelf : self
{
  public required ILogger<TSelf> Logger { protected get; set; }
  public required IStringLocalizer<TSelf> Localizer { protected get; set; }
}
class HomeController : ControllerBase<HomeController>
{
}
```

**Reply by En3Tho** [reply]
I'd rather just see potential self thingy used as 
``` csharp
record HomeController (ILogger<self> logger, IStringLocalizer<self>)...
```
Or alternatively HomeController could be just named "HC" or just "H" to save even more typing :D
